File: packages/prism-core/src/layer1/object-model/registry.ts
/**
 * ObjectRegistry — runtime registry of entity/edge type definitions, category
 * rules, and Lens slot registrations.
 *
 * The registry is the single source of truth for:
 *   - What entity types exist (registered by the Lens or application)
 *   - What each type looks like (label, icon, color, tabs, fields)
 *   - What containment is valid (category rules + per-type overrides)
 *   - What Lens slots have been registered (view panels + field extensions)
 *
 * No type strings are hardcoded here. All containment logic is derived
 * from what the consumer has registered.
 */

import type {
  GraphObject,
  EntityDef,
  EdgeTypeDef,
  EntityFieldDef,
  CategoryRule,
  TabDefinition,
} from "./types.js";

// ── Slot types ─────────────────────────────────────────────────────────────────

/**
 * A Lens's contribution to an existing entity type.
 * Slots are additive — they extend without modifying the base EntityDef.
 *
 * Field IDs contributed by a Lens should be namespaced to avoid collisions:
 *   'kami_brain_state', 'crm_deal_stage', 'palette_inventory_capacity'
 */
export interface SlotDef {
  /**
   * Unique slot identifier — use 'lensId:slotName' convention.
   * e.g. 'kami:brain', 'palette:inventory', 'crm:pipeline'
   */
  id: string;

  description?: string;

  /** Tabs/panels this slot contributes to the entity's detail view */
  tabs?: TabDefinition[];

  /**
   * Fields this slot contributes to the entity's payload schema.
   * Use namespaced IDs to prevent collisions with base fields and other slots.
   */
  fields?: EntityFieldDef[];
}

/**
 * Registers a SlotDef against one or more entity types / categories.
 * Both forTypes and forCategories are OR'd — a slot applies to any type
 * that matches either list.
 */
export interface SlotRegistration {
  slot: SlotDef;
  /** Specific entity types this slot applies to */
  forTypes?: string[];
  /** Entity categories whose members this slot applies to */
  forCategories?: string[];
}

// ── TreeNode ───────────────────────────────────────────────────────────────────

export interface TreeNode {
  object: GraphObject;
  children: TreeNode[];
  depth?: number;

  /**
   * Virtual children — objects from OTHER Lenses that reference this object.
   * Populated by WeakRefEngine, not by tree hierarchy.
   * These are read-only: cannot be moved, reparented, or deleted from here.
   */
  weakRefChildren?: WeakRefChildNode[];
}

/**
 * A virtual child node in the tree, derived from a weak-ref edge.
 * Displayed with distinct styling (dimmed, dashed border, link icon).
 * Non-draggable, non-reorderable.
 */
export interface WeakRefChildNode {
  /** The foreign object that references the parent. */
  object: GraphObject;
  /** Edge relation label (e.g. 'speaks-in', 'referenced-by'). */
  relation: string;
  /** Edge ID in the EdgeModel. */
  edgeId: string;
  /** Provider that created this ref (e.g. 'loom'). */
  providerId: string;
  /** Provider display label (e.g. 'Loom'). */
  providerLabel: string;
}

// ── ObjectRegistry ─────────────────────────────────────────────────────────────

export class ObjectRegistry<TIcon = unknown> {
  private readonly types = new Map<string, EntityDef<TIcon>>();
  private readonly rules = new Map<string, CategoryRule>();
  private readonly edgeTypes = new Map<string, EdgeTypeDef>();
  private readonly slots: SlotRegistration[] = [];

  constructor(categoryRules: CategoryRule[] = []) {
    for (const rule of categoryRules) {
      this.rules.set(rule.category, rule);
    }
  }

  // ── Entity Type Registration ─────────────────────────────────────────────────

  register(def: EntityDef<TIcon>): this {
    this.types.set(def.type, def);
    return this;
  }

  registerAll(defs: EntityDef<TIcon>[]): this {
    for (const def of defs) this.register(def);
    return this;
  }

  addCategoryRule(rule: CategoryRule): this {
    this.rules.set(rule.category, rule);
    return this;
  }

  // ── Slot Registration ────────────────────────────────────────────────────────

  /**
   * Register a Lens slot that extends one or more entity types.
   * Slots are additive — they never modify the base EntityDef.
   */
  registerSlot(registration: SlotRegistration): this {
    this.slots.push(registration);
    return this;
  }

  /**
   * All slot definitions that apply to a given entity type.
   * A slot applies if the type matches forTypes OR if the type's category
   * matches forCategories.
   */
  getSlots(type: string): SlotDef[] {
    const category = this.getCategory(type);
    return this.slots
      .filter(
        (r) =>
          (r.forTypes?.includes(type) ?? false) ||
          (r.forCategories?.includes(category) ?? false),
      )
      .map((r) => r.slot);
  }

  /**
   * Effective tabs for a type = base EntityDef tabs + all slot-contributed tabs.
   * Base tabs win on id collision (slots cannot override base tabs).
   * Falls back to a minimal default if neither the EntityDef nor any slot defines tabs.
   */
  getEffectiveTabs(type: string): TabDefinition[] {
    const base = this.types.get(type)?.tabs ?? [
      { id: "overview", label: "Overview" },
      { id: "children", label: "Children", dynamic: true },
      { id: "linked", label: "Linked", dynamic: true },
    ];
    const slotTabs = this.getSlots(type).flatMap((s) => s.tabs ?? []);
    const seen = new Set(base.map((t) => t.id));
    const merged = [...base];
    for (const t of slotTabs) {
      if (!seen.has(t.id)) {
        seen.add(t.id);
        merged.push(t);
      }
    }
    return merged;
  }

  /**
   * All field definitions for a type = EntityDef.fields + slot-contributed fields.
   * Base fields win on id collision (slots cannot override base fields).
   */
  getEntityFields(type: string): EntityFieldDef[] {
    const base = this.types.get(type)?.fields ?? [];
    const slotFields = this.getSlots(type).flatMap((s) => s.fields ?? []);
    const seen = new Set(base.map((f) => f.id));
    const merged = [...base];
    for (const f of slotFields) {
      if (!seen.has(f.id)) {
        seen.add(f.id);
        merged.push(f);
      }
    }
    return merged;
  }

  // ── Edge Type Registration ───────────────────────────────────────────────────

  registerEdge(def: EdgeTypeDef): this {
    this.edgeTypes.set(def.relation, def);
    return this;
  }

  registerEdges(defs: EdgeTypeDef[]): this {
    for (const def of defs) this.registerEdge(def);
    return this;
  }

  getEdgeType(relation: string): EdgeTypeDef | undefined {
    return this.edgeTypes.get(relation);
  }

  getEdgeLabel(relation: string): string {
    return this.edgeTypes.get(relation)?.label ?? relation;
  }

  allEdgeTypes(): string[] {
    return [...this.edgeTypes.keys()];
  }

  allEdgeDefs(): EdgeTypeDef[] {
    return [...this.edgeTypes.values()];
  }

  /**
   * Can an edge of `relation` type go from an object of `sourceType`
   * to an object of `targetType`?
   *
   * If no edge type is registered, any connection is allowed.
   */
  canConnect(
    relation: string,
    sourceType: string,
    targetType: string,
  ): boolean {
    const edgeDef = this.edgeTypes.get(relation);
    if (!edgeDef) return true;

    if (edgeDef.sourceTypes || edgeDef.sourceCategories) {
      const srcCat = this.getCategory(sourceType);
      const byType = edgeDef.sourceTypes?.includes(sourceType) ?? false;
      const byCat = edgeDef.sourceCategories?.includes(srcCat) ?? false;
      if (!byType && !byCat) return false;
    }

    if (edgeDef.targetTypes || edgeDef.targetCategories) {
      const tgtCat = this.getCategory(targetType);
      const byType = edgeDef.targetTypes?.includes(targetType) ?? false;
      const byCat = edgeDef.targetCategories?.includes(tgtCat) ?? false;
      if (!byType && !byCat) return false;
    }

    return true;
  }

  getEdgesFrom(sourceType: string): EdgeTypeDef[] {
    const cat = this.getCategory(sourceType);
    return this.allEdgeDefs().filter((def) => {
      if (!def.sourceTypes && !def.sourceCategories) return true;
      return (
        (def.sourceTypes?.includes(sourceType) ?? false) ||
        (def.sourceCategories?.includes(cat) ?? false)
      );
    });
  }

  getEdgesTo(targetType: string): EdgeTypeDef[] {
    const cat = this.getCategory(targetType);
    return this.allEdgeDefs().filter((def) => {
      if (!def.targetTypes && !def.targetCategories) return true;
      return (
        (def.targetTypes?.includes(targetType) ?? false) ||
        (def.targetCategories?.includes(cat) ?? false)
      );
    });
  }

  getEdgesBetween(sourceType: string, targetType: string): EdgeTypeDef[] {
    return this.allEdgeDefs().filter((def) =>
      this.canConnect(def.relation, sourceType, targetType),
    );
  }

  // ── Lookup ───────────────────────────────────────────────────────────────────

  get(type: string): EntityDef<TIcon> | undefined {
    return this.types.get(type);
  }

  has(type: string): boolean {
    return this.types.has(type);
  }

  allTypes(): string[] {
    return [...this.types.keys()];
  }

  allDefs(): EntityDef<TIcon>[] {
    return [...this.types.values()];
  }

  getLabel(type: string): string {
    return this.types.get(type)?.label ?? type;
  }

  getPluralLabel(type: string): string {
    const def = this.types.get(type);
    return def?.pluralLabel ?? def?.label ?? type;
  }

  getColor(type: string): string {
    return this.types.get(type)?.color ?? "#888888";
  }

  getIcon(type: string): TIcon | undefined {
    return this.types.get(type)?.icon;
  }

  getCategory(type: string): string {
    return this.types.get(type)?.category ?? "";
  }

  /**
   * Base tabs only — does not include slot contributions.
   * Use getEffectiveTabs() to get the full merged tab list.
   */
  getTabs(type: string): TabDefinition[] {
    return this.types.get(type)?.tabs ?? [
      { id: "overview", label: "Overview" },
      { id: "children", label: "Children", dynamic: true },
      { id: "linked", label: "Linked", dynamic: true },
    ];
  }

  // ── Containment ──────────────────────────────────────────────────────────────

  /**
   * Can an object of `childType` be placed inside `parentType`?
   *
   * Resolution order:
   *   1. parent.extraChildTypes includes childType -> allowed
   *   2. child.extraParentTypes includes parentType -> allowed
   *   3. parent category rule canParent includes child category -> allowed
   *   4. Otherwise -> denied
   */
  canBeChildOf(childType: string, parentType: string): boolean {
    const child = this.types.get(childType);
    const parent = this.types.get(parentType);
    if (!child || !parent) return false;

    if (parent.extraChildTypes?.includes(childType)) return true;
    if (child.extraParentTypes?.includes(parentType)) return true;

    const rule = this.rules.get(parent.category);
    if (rule?.canParent.includes(child.category)) return true;

    return false;
  }

  /**
   * Can this type exist at root level (no parent)?
   * childOnly types are always false.
   * Category canBeRoot: false overrides to false.
   * Default: true.
   */
  canBeRoot(type: string): boolean {
    const def = this.types.get(type);
    if (!def) return false;
    if (def.childOnly) return false;
    const rule = this.rules.get(def.category);
    if (rule?.canBeRoot === false) return false;
    return true;
  }

  /**
   * Can this type have children?
   * true if category rule has at least one canParent entry,
   * or if the type has extraChildTypes defined.
   */
  canHaveChildren(type: string): boolean {
    const def = this.types.get(type);
    if (!def) return false;
    if (def.childOnly) return false;
    if (def.extraChildTypes && def.extraChildTypes.length > 0) return true;
    const rule = this.rules.get(def.category);
    return (rule?.canParent.length ?? 0) > 0;
  }

  /** All registered types that can be placed inside `parentType`. */
  getAllowedChildTypes(parentType: string): string[] {
    return this.allTypes().filter((t) => this.canBeChildOf(t, parentType));
  }

  // ── Tree Utilities ───────────────────────────────────────────────────────────

  buildTree(objects: GraphObject[]): TreeNode[] {
    const nodeMap = new Map<string, TreeNode>();
    for (const obj of objects) {
      nodeMap.set(obj.id, { object: obj, children: [] });
    }
    const roots: TreeNode[] = [];
    for (const node of nodeMap.values()) {
      const parent = node.object.parentId
        ? nodeMap.get(node.object.parentId)
        : null;
      if (parent) {
        parent.children.push(node);
      } else {
        roots.push(node);
      }
    }
    const sortByPosition = (nodes: TreeNode[]): void => {
      nodes.sort((a, b) => a.object.position - b.object.position);
      for (const n of nodes) sortByPosition(n.children);
    };
    sortByPosition(roots);
    return roots;
  }

  getAncestors(
    id: string,
    objectMap: Map<string, GraphObject>,
  ): GraphObject[] {
    const ancestors: GraphObject[] = [];
    let current = objectMap.get(id);
    while (current?.parentId) {
      const parent = objectMap.get(current.parentId);
      if (!parent) break;
      ancestors.push(parent);
      current = parent;
    }
    return ancestors;
  }
}

File: packages/prism-core/src/foundation/object_model/registry.rs
//! `ObjectRegistry` — runtime registry of entity/edge type
//! definitions, category rules, and Lens slot registrations. Port
//! of `foundation/object-model/registry.ts`.
//!
//! Unlike the TS original the Rust port does not parameterize over
//! an icon type. Icons are just strings (icon name / SVG id) and
//! any host that needs richer representations can keep a parallel
//! table keyed by type.

use std::collections::HashMap;

use indexmap::IndexMap;
use serde::{Deserialize, Serialize};

use super::types::{
    CategoryRule, EdgeTypeDef, EntityDef, EntityFieldDef, GraphObject, TabDefinition,
};

// ── Slot types ─────────────────────────────────────────────────────

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SlotDef {
    pub id: String,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub tabs: Option<Vec<TabDefinition>>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub fields: Option<Vec<EntityFieldDef>>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SlotRegistration {
    pub slot: SlotDef,
    #[serde(skip_serializing_if = "Option::is_none", rename = "forTypes", default)]
    pub for_types: Option<Vec<String>>,
    #[serde(
        skip_serializing_if = "Option::is_none",
        rename = "forCategories",
        default
    )]
    pub for_categories: Option<Vec<String>>,
}

// ── TreeNode ───────────────────────────────────────────────────────

#[derive(Debug, Clone, PartialEq)]
pub struct TreeNode {
    pub object: GraphObject,
    pub children: Vec<TreeNode>,
    pub depth: Option<usize>,
    pub weak_ref_children: Option<Vec<WeakRefChildNode>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct WeakRefChildNode {
    pub object: GraphObject,
    pub relation: String,
    pub edge_id: String,
    pub provider_id: String,
    pub provider_label: String,
}

// ── ObjectRegistry ─────────────────────────────────────────────────

/// Runtime registry of entity types, edge types, category rules
/// and lens slots. `IndexMap` is used for deterministic iteration
/// order to match the legacy JS `Map` insertion-order semantics.
#[derive(Debug, Default, Clone)]
pub struct ObjectRegistry {
    types: IndexMap<String, EntityDef>,
    rules: HashMap<String, CategoryRule>,
    edge_types: IndexMap<String, EdgeTypeDef>,
    slots: Vec<SlotRegistration>,
}

impl ObjectRegistry {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn with_category_rules(rules: impl IntoIterator<Item = CategoryRule>) -> Self {
        let mut reg = Self::default();
        for rule in rules {
            reg.rules.insert(rule.category.clone(), rule);
        }
        reg
    }

    // ── Entity Type Registration ──────────────────────────────────

    pub fn register(&mut self, def: EntityDef) -> &mut Self {
        self.types.insert(def.type_name.clone(), def);
        self
    }

    pub fn register_all<I>(&mut self, defs: I) -> &mut Self
    where
        I: IntoIterator<Item = EntityDef>,
    {
        for def in defs {
            self.register(def);
        }
        self
    }

    pub fn remove(&mut self, type_name: &str) -> bool {
        self.types.shift_remove(type_name).is_some()
    }

    pub fn add_category_rule(&mut self, rule: CategoryRule) -> &mut Self {
        self.rules.insert(rule.category.clone(), rule);
        self
    }

    // ── Slot Registration ─────────────────────────────────────────

    pub fn register_slot(&mut self, registration: SlotRegistration) -> &mut Self {
        self.slots.push(registration);
        self
    }

    pub fn get_slots(&self, type_name: &str) -> Vec<&SlotDef> {
        let category = self.get_category(type_name);
        self.slots
            .iter()
            .filter(|reg| {
                let matches_type = reg
                    .for_types
                    .as_ref()
                    .map(|ts| ts.iter().any(|t| t == type_name))
                    .unwrap_or(false);
                let matches_cat = reg
                    .for_categories
                    .as_ref()
                    .map(|cs| cs.iter().any(|c| c == category))
                    .unwrap_or(false);
                matches_type || matches_cat
            })
            .map(|reg| &reg.slot)
            .collect()
    }

    pub fn get_effective_tabs(&self, type_name: &str) -> Vec<TabDefinition> {
        let base: Vec<TabDefinition> = self
            .types
            .get(type_name)
            .and_then(|d| d.tabs.clone())
            .unwrap_or_else(default_tabs);
        let mut seen: std::collections::HashSet<String> =
            base.iter().map(|t| t.id.clone()).collect();
        let mut merged = base;
        for slot in self.get_slots(type_name) {
            if let Some(tabs) = slot.tabs.as_ref() {
                for t in tabs {
                    if !seen.contains(&t.id) {
                        seen.insert(t.id.clone());
                        merged.push(t.clone());
                    }
                }
            }
        }
        merged
    }

    pub fn get_entity_fields(&self, type_name: &str) -> Vec<EntityFieldDef> {
        let base: Vec<EntityFieldDef> = self
            .types
            .get(type_name)
            .and_then(|d| d.fields.clone())
            .unwrap_or_default();
        let mut seen: std::collections::HashSet<String> =
            base.iter().map(|f| f.id.clone()).collect();
        let mut merged = base;
        for slot in self.get_slots(type_name) {
            if let Some(fields) = slot.fields.as_ref() {
                for f in fields {
                    if !seen.contains(&f.id) {
                        seen.insert(f.id.clone());
                        merged.push(f.clone());
                    }
                }
            }
        }
        merged
    }

    // ── Edge Type Registration ────────────────────────────────────

    pub fn register_edge(&mut self, def: EdgeTypeDef) -> &mut Self {
        self.edge_types.insert(def.relation.clone(), def);
        self
    }

    pub fn register_edges<I>(&mut self, defs: I) -> &mut Self
    where
        I: IntoIterator<Item = EdgeTypeDef>,
    {
        for def in defs {
            self.register_edge(def);
        }
        self
    }

    pub fn remove_edge(&mut self, relation: &str) -> bool {
        self.edge_types.shift_remove(relation).is_some()
    }

    pub fn get_edge_type(&self, relation: &str) -> Option<&EdgeTypeDef> {
        self.edge_types.get(relation)
    }

    pub fn get_edge_label<'a>(&'a self, relation: &'a str) -> &'a str {
        self.edge_types
            .get(relation)
            .map(|d| d.label.as_str())
            .unwrap_or(relation)
    }

    pub fn all_edge_types(&self) -> Vec<&str> {
        self.edge_types.keys().map(String::as_str).collect()
    }

    pub fn all_edge_defs(&self) -> Vec<&EdgeTypeDef> {
        self.edge_types.values().collect()
    }

    pub fn can_connect(&self, relation: &str, source_type: &str, target_type: &str) -> bool {
        let Some(def) = self.edge_types.get(relation) else {
            return true;
        };

        if def.source_types.is_some() || def.source_categories.is_some() {
            let src_cat = self.get_category(source_type);
            let by_type = def
                .source_types
                .as_ref()
                .map(|ts| ts.iter().any(|t| t == source_type))
                .unwrap_or(false);
            let by_cat = def
                .source_categories
                .as_ref()
                .map(|cs| cs.iter().any(|c| c == src_cat))
                .unwrap_or(false);
            if !by_type && !by_cat {
                return false;
            }
        }

        if def.target_types.is_some() || def.target_categories.is_some() {
            let tgt_cat = self.get_category(target_type);
            let by_type = def
                .target_types
                .as_ref()
                .map(|ts| ts.iter().any(|t| t == target_type))
                .unwrap_or(false);
            let by_cat = def
                .target_categories
                .as_ref()
                .map(|cs| cs.iter().any(|c| c == tgt_cat))
                .unwrap_or(false);
            if !by_type && !by_cat {
                return false;
            }
        }

        true
    }

    pub fn get_edges_from(&self, source_type: &str) -> Vec<&EdgeTypeDef> {
        let cat = self.get_category(source_type).to_string();
        self.all_edge_defs()
            .into_iter()
            .filter(|def| {
                if def.source_types.is_none() && def.source_categories.is_none() {
                    return true;
                }
                let by_type = def
                    .source_types
                    .as_ref()
                    .map(|ts| ts.iter().any(|t| t == source_type))
                    .unwrap_or(false);
                let by_cat = def
                    .source_categories
                    .as_ref()
                    .map(|cs| cs.iter().any(|c| c == &cat))
                    .unwrap_or(false);
                by_type || by_cat
            })
            .collect()
    }

    pub fn get_edges_to(&self, target_type: &str) -> Vec<&EdgeTypeDef> {
        let cat = self.get_category(target_type).to_string();
        self.all_edge_defs()
            .into_iter()
            .filter(|def| {
                if def.target_types.is_none() && def.target_categories.is_none() {
                    return true;
                }
                let by_type = def
                    .target_types
                    .as_ref()
                    .map(|ts| ts.iter().any(|t| t == target_type))
                    .unwrap_or(false);
                let by_cat = def
                    .target_categories
                    .as_ref()
                    .map(|cs| cs.iter().any(|c| c == &cat))
                    .unwrap_or(false);
                by_type || by_cat
            })
            .collect()
    }

    pub fn get_edges_between(&self, source_type: &str, target_type: &str) -> Vec<&EdgeTypeDef> {
        let relations: Vec<String> = self
            .edge_types
            .values()
            .filter(|def| self.can_connect(&def.relation, source_type, target_type))
            .map(|def| def.relation.clone())
            .collect();
        relations
            .into_iter()
            .filter_map(|r| self.edge_types.get(&r))
            .collect()
    }

    // ── Lookup ────────────────────────────────────────────────────

    pub fn get(&self, type_name: &str) -> Option<&EntityDef> {
        self.types.get(type_name)
    }

    pub fn has(&self, type_name: &str) -> bool {
        self.types.contains_key(type_name)
    }

    pub fn all_types(&self) -> Vec<&str> {
        self.types.keys().map(String::as_str).collect()
    }

    pub fn all_defs(&self) -> Vec<&EntityDef> {
        self.types.values().collect()
    }

    pub fn get_label<'a>(&'a self, type_name: &'a str) -> &'a str {
        self.types
            .get(type_name)
            .map(|d| d.label.as_str())
            .unwrap_or(type_name)
    }

    pub fn get_plural_label<'a>(&'a self, type_name: &'a str) -> &'a str {
        let def = self.types.get(type_name);
        def.and_then(|d| d.plural_label.as_deref())
            .or_else(|| def.map(|d| d.label.as_str()))
            .unwrap_or(type_name)
    }

    pub fn get_color(&self, type_name: &str) -> &str {
        self.types
            .get(type_name)
            .and_then(|d| d.color.as_deref())
            .unwrap_or("#888888")
    }

    pub fn get_category(&self, type_name: &str) -> &str {
        self.types
            .get(type_name)
            .map(|d| d.category.as_str())
            .unwrap_or("")
    }

    pub fn get_tabs(&self, type_name: &str) -> Vec<TabDefinition> {
        self.types
            .get(type_name)
            .and_then(|d| d.tabs.clone())
            .unwrap_or_else(default_tabs)
    }

    // ── Containment ───────────────────────────────────────────────

    pub fn can_be_child_of(&self, child_type: &str, parent_type: &str) -> bool {
        let Some(child) = self.types.get(child_type) else {
            return false;
        };
        let Some(parent) = self.types.get(parent_type) else {
            return false;
        };

        if let Some(extras) = parent.extra_child_types.as_ref() {
            if extras.iter().any(|t| t == child_type) {
                return true;
            }
        }
        if let Some(extras) = child.extra_parent_types.as_ref() {
            if extras.iter().any(|t| t == parent_type) {
                return true;
            }
        }

        if let Some(rule) = self.rules.get(&parent.category) {
            if rule.can_parent.iter().any(|c| c == &child.category) {
                return true;
            }
        }

        false
    }

    pub fn can_be_root(&self, type_name: &str) -> bool {
        let Some(def) = self.types.get(type_name) else {
            return false;
        };
        if def.child_only.unwrap_or(false) {
            return false;
        }
        if let Some(rule) = self.rules.get(&def.category) {
            if rule.can_be_root == Some(false) {
                return false;
            }
        }
        true
    }

    pub fn can_have_children(&self, type_name: &str) -> bool {
        let Some(def) = self.types.get(type_name) else {
            return false;
        };
        if def.child_only.unwrap_or(false) {
            return false;
        }
        if def
            .extra_child_types
            .as_ref()
            .map(|v| !v.is_empty())
            .unwrap_or(false)
        {
            return true;
        }
        self.rules
            .get(&def.category)
            .map(|r| !r.can_parent.is_empty())
            .unwrap_or(false)
    }

    pub fn get_allowed_child_types(&self, parent_type: &str) -> Vec<String> {
        self.types
            .keys()
            .filter(|t| self.can_be_child_of(t, parent_type))
            .cloned()
            .collect()
    }

    // ── Tree Utilities ────────────────────────────────────────────

    pub fn build_tree(&self, objects: &[GraphObject]) -> Vec<TreeNode> {
        // Two-pass assembly: first index every object by id, then
        // recursively materialise each subtree starting from the
        // detected roots. This avoids the order-sensitivity that
        // bit the earlier single-pass version.
        let by_id: HashMap<&str, &GraphObject> =
            objects.iter().map(|o| (o.id.as_str(), o)).collect();

        let mut children_of: HashMap<&str, Vec<&GraphObject>> = HashMap::new();
        let mut root_ids: Vec<&str> = Vec::new();
        for obj in objects {
            match obj.parent_id.as_ref().map(|p| p.as_str()) {
                Some(pid) if by_id.contains_key(pid) => {
                    children_of.entry(pid).or_default().push(obj);
                }
                _ => root_ids.push(obj.id.as_str()),
            }
        }

        fn build<'a>(
            obj: &'a GraphObject,
            children_of: &HashMap<&'a str, Vec<&'a GraphObject>>,
        ) -> TreeNode {
            let kids = children_of
                .get(obj.id.as_str())
                .map(|v| v.iter().map(|o| build(o, children_of)).collect())
                .unwrap_or_default();
            TreeNode {
                object: obj.clone(),
                children: kids,
                depth: None,
                weak_ref_children: None,
            }
        }

        let mut roots: Vec<TreeNode> = root_ids
            .into_iter()
            .filter_map(|id| by_id.get(id).map(|obj| build(obj, &children_of)))
            .collect();

        sort_by_position(&mut roots);
        roots
    }

    pub fn get_ancestors(
        &self,
        id: &str,
        object_map: &HashMap<String, GraphObject>,
    ) -> Vec<GraphObject> {
        let mut ancestors = Vec::new();
        let mut current = object_map.get(id);
        while let Some(obj) = current {
            if let Some(parent_id) = obj.parent_id.as_ref() {
                if let Some(parent) = object_map.get(parent_id.as_str()) {
                    ancestors.push(parent.clone());
                    current = Some(parent);
                    continue;
                }
            }
            break;
        }
        ancestors
    }
}

fn default_tabs() -> Vec<TabDefinition> {
    vec![
        TabDefinition {
            id: "overview".into(),
            label: "Overview".into(),
            dynamic: None,
        },
        TabDefinition {
            id: "children".into(),
            label: "Children".into(),
            dynamic: Some(true),
        },
        TabDefinition {
            id: "linked".into(),
            label: "Linked".into(),
            dynamic: Some(true),
        },
    ]
}

fn sort_by_position(nodes: &mut Vec<TreeNode>) {
    nodes.sort_by(|a, b| {
        a.object
            .position
            .partial_cmp(&b.object.position)
            .unwrap_or(std::cmp::Ordering::Equal)
    });
    for n in nodes {
        sort_by_position(&mut n.children);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::foundation::object_model::types::{DefaultChildView, ObjectId};

    fn task_def() -> EntityDef {
        EntityDef {
            type_name: "task".into(),
            nsid: None,
            category: "record".into(),
            label: "Task".into(),
            plural_label: Some("Tasks".into()),
            description: None,
            color: Some("#ff0000".into()),
            default_child_view: Some(DefaultChildView::List),
            tabs: None,
            child_only: None,
            extra_child_types: None,
            extra_parent_types: None,
            fields: None,
            api: None,
        }
    }

    fn project_def() -> EntityDef {
        EntityDef {
            type_name: "project".into(),
            nsid: None,
            category: "container".into(),
            label: "Project".into(),
            plural_label: Some("Projects".into()),
            description: None,
            color: None,
            default_child_view: None,
            tabs: None,
            child_only: None,
            extra_child_types: None,
            extra_parent_types: None,
            fields: None,
            api: None,
        }
    }

    fn container_rule() -> CategoryRule {
        CategoryRule {
            category: "container".into(),
            can_parent: vec!["container".into(), "record".into()],
            can_be_root: Some(true),
        }
    }

    fn record_rule() -> CategoryRule {
        CategoryRule {
            category: "record".into(),
            can_parent: Vec::new(),
            can_be_root: Some(true),
        }
    }

    #[test]
    fn containment_rules_respect_categories() {
        let mut reg = ObjectRegistry::with_category_rules([container_rule(), record_rule()]);
        reg.register(project_def()).register(task_def());
        assert!(reg.can_be_child_of("task", "project"));
        assert!(!reg.can_be_child_of("project", "task"));
        assert!(reg.can_be_root("project"));
        assert!(reg.can_be_root("task"));
    }

    #[test]
    fn build_tree_sorts_by_position() {
        let reg = ObjectRegistry::new();
        let mut a = GraphObject::new("a", "project", "A");
        a.position = 1.0;
        let mut b = GraphObject::new("b", "task", "B");
        b.parent_id = Some(ObjectId::new("a"));
        b.position = 0.0;
        let mut c = GraphObject::new("c", "task", "C");
        c.parent_id = Some(ObjectId::new("a"));
        c.position = 1.0;
        let tree = reg.build_tree(&[c.clone(), a.clone(), b.clone()]);
        assert_eq!(tree.len(), 1);
        assert_eq!(tree[0].object.id.as_str(), "a");
        assert_eq!(tree[0].children[0].object.id.as_str(), "b");
        assert_eq!(tree[0].children[1].object.id.as_str(), "c");
    }
}
